### PR TITLE
Implement (most of) Ineluki Key Patch

### LINKED ISSUE #1181: Ineluki Key Patch
DL: http://cherrytree.at/cms/download/?did=3

"This famous tool (which isn’t really a patch but a DLL replacement) allows full keyboard and mouse input as well es executing external programs from an RPG Maker event. It also provides MP3 support, which is very complicated and buggy, however."

Basicly a harmony.dll replacement with "scripts" written as INI files in Music folder...

Files are of format ".script.wav". Player already detects them and reports "Nope!"

### PR DESCRIPTION
**Pull request by Ghabry**
Can we agree on that this fixes #1181 and  #1367 even though there are more problems here?
Imo there can be new issues opened when they occur.

For #1181 not all ExecProgram calls are supported.
#1367 also has a patches exe to skip the title screen and load the first save, not related to Ineluki.

Limitations:
- ``ExecProgram`` not supported except for some specific programs that are HLE'ed
- ``MciProgram`` not supported
- Only alphanumeric and some special keys supported. This doesn't map properly to the key codes because Ineluki was written with a German keyboard in mind
- Cheat support does not reset when wrong key is pressed, needs some kind of input callback API, polling is ugly and highly inefficient here.

Problems:
- Mouse support needs ``--enable-mouse`` argument and this binds clicks and wheel to confirm and other stuff. So it conflicts with the patch
- WSAD conflict. We use them for walking but in RPG_RT they are not mapped.

Ideas are welcome about how to resolve the problems.

Tests:
- https://easyrpg.org/play/pr2397/?game=ineluki&enable-mouse
- https://easyrpg.org/play/pr2397/?game=ineluki_autorun&enable-mouse - Tests that the autorun.script is executed on startup
- https://easyrpg.org/play/pr2397/?game=calm_falls MP3 link file support in the web player (vs. before https://easyrpg.org/play/?game=calm_falls )

Best for testing are the two samples at the bottom:
- Diablo-Like Mouse control (click to walk)
- Tasten Test (Key Test) - Prints the Key ID

### CODE REVIEW

**Comment by Ghabry** [comment]
Also added Ineluki MP3 Link file support to the web Player:

https://easyrpg.org/play/pr2397/?game=calm_falls

The title and the first map music are already links.

File: src/player.cpp [hunk 1]
```diff
@@ -370,7 +371,13 @@ void Player::Update(bool update_scene) {
 	}
 
 	if (update_scene) {
+		if (Main_Data::game_ineluki) {
+			Main_Data::game_ineluki->Update();
+		}
+
 		Scene::instance->Update();
+
+		IncFrame();
```
Comment (Ghabry) [inline src/player.cpp:380]: oh bad rebase, must remove this IncFrame. sry

**Comment by Ghabry** [comment]
This is ready for testing. The "Problems" should be discussed but otherwise I'm done here.

**Review by fdelapena — APPROVED** [review]

**Comment by fdelapena** [comment]
Works for me :+1:. Regarding the need of adding the mouse arg and the WASD keys conflict, perhaps a flag when ineluki's key patch is detected somehow to toggle the mouse and keys behaviour could be a possible approach to handle this.

**Comment by Ghabry** [comment]
WSAD and the mouse issue is solved:

I invented a new feature "key masking". The Input system can be told which keys are never reported as pressed.

I mask now WSAD or the Left/Right Mouse when the key/mouse support of Ineluki is requested.

This also allows now to always process mouse events in SDL2: When the ``mouse_flag (--enable-mouse)`` is not set then the input system masks the mouse now instead. (The ``mouse_flag`` maps left click to Confirm and scroll to Page up/down)

**Pull request merged**